### PR TITLE
Fix child element indentation in multiline lists

### LINKED ISSUE #3817: Visual Studio formatting is wrong in some cases
Here is an example

``` typescript
var webApps: ng.ui.IState[] = [{
    name: "home.webapp",
    templateUrl: "templates/empty-shell.html"
}, {
    name: "home.webapp.templates",
    templateUrl: "templates/templates.html"
}, {
    name: "home.webapp.work",
    templateUrl: "templates/work.html"
}];
```

VS formats it to

``` typescript
var webApps: ng.ui.IState[] = [{
    name: "home.webapp",
    templateUrl: "templates/empty-shell.html"
}, {
        name: "home.webapp.templates",
        templateUrl: "templates/templates.html"
    }, {
        name: "home.webapp.work",
        templateUrl: "templates/work.html"
    }];
```

another example is 

``` typescript
function test($http: ng.IHttpService) {
    $http({
        url: "https://github.com",
        method: "GET"
    })
    .success((d) => {
        console.log(d);
    })
    .finally(() => {
        console.log("finally");
    });
}
```

VS formats it to 

``` typescript
function test($http: ng.IHttpService) {
    $http({
        url: "https://github.com",
        method: "GET"
    })
        .success((d) => {
            console.log(d);
        })
        .finally(() => {
            console.log("finally");
        });
}
```

### LINKED ISSUE #5834: Keep indentation level consistent in a multiline list
Copied from #3436:

> Another example:
> 
> ``` typescript
> return (
>     h.div({
>         className: "view-node-row",
>         style: {
>             width: width,
>             height: rowHeight
>         }
>     },
>         backgroundElement,
>         centerElement,
>         toolsElement,
>         h.div({ className: "nodes" }, items)
>         )
>     );
> ```
> 
> While I expect:
> 
> ``` typescript
> return (
>     h.div({
>         className: "view-node-row",
>         style: {
>             width: width,
>             height: rowHeight
>         }
>     },
>     backgroundElement,
>     centerElement,
>     toolsElement,
>     h.div({ className: "nodes" }, items)
>     )
> );
> ```

### LINKED ISSUE #9437: Indentation bug on array
**TypeScript Version:**  master branch

**Code**

Copy-pasting #7503 but with a slight change...

```ts
export let Things = [3, {
    Hat: 'hat',
    Glove: 'glove',
    Umbrella: 'umbrella'
},
    {
        Salad: 'salad',
        Burrito: 'burrito',
        Pie: 'pie'
    }];
```

**Expected behavior:**

```ts
export let Things = [3, {
    Hat: 'hat',
    Glove: 'glove',
    Umbrella: 'umbrella'
},
{
    Salad: 'salad',
    Burrito: 'burrito',
    Pie: 'pie'
}];
```

**Actual behavior:** No change

### PR DESCRIPTION
**Pull request by SaschaNaz**
Fixes #3817, fixes #5834, fixes #9437.

Please don't do commit-by-commit review as some intermediate changes were later removed.

**Note (July 1, 2016):** This sadly reverts #8364.

Input: 

``` typescript
return (
    h.div({
    },
        ""
    )
);

[3, {
}, {
    }]
```

Fix: 

``` typescript
return (
    h.div({
    },
    ""
    )
);

[3, {
}, {
}]
```

### CODE REVIEW

**Comment by SaschaNaz** [comment]
Now waiting feedback.

**Comment by wgebczyk** [comment]
So after the fix we have:
```
return (
    h.div({
        className: "view-node-row",
        style: {
            width: width,
            height: rowHeight
        }
    },
    backgroundElement,
    centerElement,
    toolsElement,
    h.div({ className: "nodes" }, items)
    )
);
```
this means probably we will have this:
```
return (
    h.div(backgroundElement,
    centerElement,
    toolsElement,
    h.div({ className: "nodes" }, items)
    )
);
```
and seems that by unwrapping return we will have this:
```
h.div(backgroundElement,
centerElement,
toolsElement,
h.div({ className: "nodes" }, items)
);
```
Is that correct?

Honestly I don't like situation when function parameters are indented at same level as its function call.
But that's my preference...

**Comment by SaschaNaz** [comment]
@wgebczyk No, this fix will affect only when:

1. The end position of a parameter is in a different line with the start position of the list
2. The start position of the parameter is in the same line with the start position of the list

```typescript
foo({ // An object literal starts in the first line
}, // An object literal ends in the second line, not the first line
3 // Thus, no additional indentation here
),

foo(
    { // starts in the second line
    },  
    3 // Additional indentation here
);

foo({}, // starts and ends in the first line
    3 // Additional indentation here
);
```

**Comment by wgebczyk** [comment]
Seems to be reasonable. Thanks!

File: src/services/formatting/formatting.ts [hunk 1]
```diff
@@ -605,22 +609,62 @@ namespace ts.formatting {
...
+                        parentDynamicIndentation = getDynamicIndentation(parent, parentStartLine, inheritedIndentation, 0);
+                    }
+
                     // if child node is a token, it does not impact indentation, proceed it using parent indentation scope rules
                     let tokenInfo = formattingScanner.readTokenInfo(child);
                     Debug.assert(tokenInfo.token.end === child.end);
                     consumeTokenAndAdvanceScanner(tokenInfo, node, parentDynamicIndentation);
-                    return inheritedIndentation;
+                    if (checkListElementLockIndentation()) {
+                        parentListStates.indentationLocked = true;
```
Comment (DanielRosenwasser) [inline src/services/formatting/formatting.ts]: Above you check for `parentListStates` being defined; but can't it be undefined here?

**Comment by mhegazy** [comment]
@vladima can you take a look?

**Comment by aleksey-bykov** [comment]
can we merge it please?

**Comment by SaschaNaz** [comment]
I'm closing this because this is incompatible with new PR #13574. I will open a cleaner PR if it is merged.

**Pull request closed without merging**